### PR TITLE
Fix #98209. Test all agda packages

### LINKED ISSUE #98209: Test whether all agda packages build on CI
In https://github.com/NixOS/nixpkgs/pull/98199 I nearly broke agda packages that depend on the agda standard library. It would be nice if there was some automatised check that builds all agda packages. Currently it's not very many packages, so I don't expect this to be a heavy load addition to CI.

I added such a functionality in this commit: https://github.com/NixOS/nixpkgs/pull/98199/commits/e1e2d431fb9fb2e58f9339200a58d5ba54c99e1e But it was out of scope for that PR, and also not a approach in general because that test doesn't concern the agda package set, but agda itself.

In this issue, we should determine and implement whether and how we test whether all agda packages build.

My proposal is adding a small test that builds all agda packages that are not marked as broken.

@alexarice

### PR DESCRIPTION
**Pull request by turion**
###### Motivation for this change
Test all agda packages

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

@alexarice This is WIP still.

### CODE REVIEW

**Comment by turion** [comment]
Currently rebased onto https://github.com/NixOS/nixpkgs/pull/98199 to get a feeling whether this is feasible.

**Comment by alexarice** [comment]
I'm not sure this achieves what you want in that I don't think these tests will be ran by ofBorg when you update standard-library

**Comment by turion** [comment]
Yes, true :/ Do you have an idea how to go about this?

**Comment by alexarice** [comment]
Disclaimer that I am not sure about this but this is what I believe happens:

- When you submit the PR, by default ofBorg just checks that the nix evaluates (i.e. doesn't build anything)
- If the user is recognised(?) by ofBorg (via putting in a pr to ofBorg) and the pr title has a certain format, or if someone types ofBorg build (but actually tagging ofBorg), it tries to build the package and runs any tests contained in `passthru.tests`

However the CI checks have looked a bit different recently, and I'm not sure if there is a detailed description of what they do somewhere

**Comment by turion** [comment]
@alexarice Does that mean this is still useful? Maybe one of us can get ofBorg rights to trigger the build.

**Comment by alexarice** [comment]
I have ofBorg rights, that is not the problem. The problem is that as I understand it you want:
PR put in for standard library -> ofBorg tests all downstream packages.

The test you have added does not do this. It adds checking all agda packages to the agda test, which can only be run manually.

**Comment by turion** [comment]
Right now it's fine for me if the packages are built either automatically or by triggering ofBorg manually. These PRs are still rare enough that we'll always notice. I'm fine with going with the solution that is most likely to be merged.

I could even imagine an `allPackages` derivation that simply has all agda packages as build inputs and does nothing otherwise, and when reviewing one can simply build that locally to verify that nothing was broken.

**Comment by alexarice** [comment]
Maybe there is a standard way to check for breakages in downstream packages?

**Comment by alexarice** [comment]
I just don't feel this test is checking if the agda infrastructure is broken which shouldn't be the case if one leaf package is a bit dodgy

**Comment by turion** [comment]
> Maybe there is a standard way to check for breakages in downstream packages?

How is it done in Haskell?

**Comment by alexarice** [comment]
They update the whole package set with cabal2nix at one time and then I guess build the whole thing manually and see what packages broke though I am not an expert

**Comment by turion** [comment]
Ok, we can't quite copy that workflow yet (see https://github.com/NixOS/nixpkgs/pull/87903 though for an initial idea how to do that). How about the easiest way, just make a set `allUnbrokenPackages`, document it, and we can remind people on PRs to build it locally in order to ensure that there are no breakages?

**Comment by alexarice** [comment]
Does anything else do this?

**Comment by turion** [comment]
> Does anything else do this?

I don't know, I'm not really involved in any other language frameworks.

**Comment by alexarice** [comment]
I'm not really against it, just seems odd to have a meta-package in the package set, really feels like it should be a test

**Comment by turion** [comment]
Understandable.

So a separate test? In `nixos/tests/agda/allUnbrokenPackages.nix`? How would it look like? It wouldn't do anything in the body, it just has all unbroken packages as dependencies?

**Comment by alexarice** [comment]
I'm afraid I don't really know enough about testing

File: pkgs/build-support/agda/default.nix [hunk 1]
```diff
@@ -73,6 +76,9 @@ let
           find \( ${concatMapStringsSep " -or " (p: "-name '*.${p}'") (extensions ++ extraExtensions)} \) -exec cp -p --parents -t "$out" {} +
           runHook postInstall
         '';
+
+        # Retrieve the current package from the finished package set and try to build it
+        passthru.tests.build = builtins.getAttr pname agdaPackages;
```
Comment (turion) [inline pkgs/build-support/agda/default.nix:81]: @alexarice How about this? If it works I can drop the other commits and leave the nixos test as it is.

**Comment by turion** [comment]
@neosimsim see this WIP

**Comment by turion** [comment]
@ofborg test agda

**Comment by turion** [comment]
@ofborg test agdaPackages.agda-categories.passthru.tests.build

**Comment by turion** [comment]
@ofborg build agdaPackages.agda-categories

**Comment by turion** [comment]
@alexarice I only want to have _either_ testing all agda libraries in the nixos test, _or_ testing each agda library in each `agdaPackages.mylib.passthru.tests.build`. What's better? I think the second one.

File: pkgs/build-support/agda/default.nix (hunk 1 above)
Comment (turion) [reply pkgs/build-support/agda/default.nix:81]: Come to think about it, I find this strange. All this does is to check whether the package builds. That should be covered by saying "ofborg build agdaPackages.mylibrary" shouldn't it?

**Comment by alexarice** [comment]
> 
> 
> @alexarice I only want to have _either_ testing all agda libraries in the nixos test, _or_ testing each agda library in each `agdaPackages.mylib.passthru.tests.build`. What's better? I think the second one.

The second one makes sense to me. In one of the recent PRs, it seemed like ofborg tested downsteam packages automatically? or is this me imagining things

**Comment by turion** [comment]
What do you mean by downstream package? When it builds e.g. `agda-categories`, it needs to build all its dependencies anyways.

Huh. I'm realising that what I have done in `passthru.tests` doesn't make sense. What I really need to test are the _reverse_ dependencies, because those aren't built automatically! Maybe that's what you're asking?

**Comment by alexarice** [comment]
Sorry I did mean downstream dependencies

**Comment by turion** [comment]
I'm not sure what that means in this context. Do you consider e.g. `agda-categories` downstream of `agda-packages`?

**Comment by alexarice** [comment]
agda-categories would be downstream of standard-library

**Comment by turion** [comment]
Ok. Then we could either look into a way to generate a list of these downstream/reverse dependent packages and build only those in the test, or we could build all packages in the test. The latter is simpler, but more wasteful, and also annoying if an unrelated package is already broken. So I'll try to do the former, if you think that's useful.

**Comment by turion** [comment]
@ofborg build agdaPackages.agda-categories
@ofborg build agdaPackages.standard-library

**Comment by turion** [comment]
@ofborg build agdaPackages.agda-categories
@ofborg build agdaPackages.standard-library

**Comment by turion** [comment]
The package itself fails to build (I added a build failure), but the tests don't seem to be run.

**Comment by turion** [comment]
@ofborg build agdaPackages.agda-categories
@ofborg build agdaPackages.standard-library

**Comment by turion** [comment]
Compare e.g. locally:

```
$ nix repl .
Welcome to Nix version 2.3.8. Type :? for help.

Loading '.'...
Added 13288 variables.

nix-repl> :b agdaPackages.standard-library.passthru.tests.agda-categories
builder for '/nix/store/3s95n2jfiknz0h22ha7pjaascar34wm6-agda-categories-0.1.4.drv' failed with exit code 1; last 7 log lines:
  unpacking sources
  unpacking source archive /nix/store/bgzixb5wbxbivc7aznhyjzb2pc66ysnf-source
  source root is source
  patching sources
  configuring
  no configure script, doing nothing
  building
[0 built (1 failed)]
error: build of '/nix/store/3s95n2jfiknz0h22ha7pjaascar34wm6-agda-categories-0.1.4.drv' failed

```

**Comment by turion** [comment]
@ofborg build agdaPackages.standard-library.passthru.tests

**Comment by turion** [comment]
@alexarice @neosimsim I think this is in good shape now. Feel free to review :)

**Comment by turion** [comment]
I've also added guidelines on how to maintain the agda libraries. Opinions welcome :)

**Comment by turion** [comment]
@ofborg build agdaPackages

**Comment by turion** [comment]
It would still be good to have a test for _all_ agda packages, not just each reverse dependency, in case we change something in the build system.

**Comment by neosimsim** [comment]
> It would still be good to have a test for _all_ agda packages, not just each reverse dependency, in case we change something in the build system.

Would it suffice to treat packages depending on `Adga` as "all agda packages".

**Comment by turion** [comment]
Not necessarily, other packages in nixpkgs might in theory depend on agda (although I doubt any currently exist).

**Comment by alexarice** [comment]
All agda packages currently have `isAgdaDerivation = true` so that can be used. I wouldn't mind putting a "buildAll" package in `agda.passthru.tests`.

**Comment by turion** [comment]
> I wouldn't mind putting a "buildAll" package in agda.passthru.tests.

That sounds good. How would I make one? I.e. how do I combine all the derivations to one?

**Comment by alexarice** [comment]
I'm using this,

```
all = withPackages (builtins.filter (x: x ? isAgdaDerivation) (lib.attrValues (removeAttrs self ["all"])));
```

**Comment by turion** [comment]
Thanks, I gave it a shot. I'm not sure whether that's the right place for the recursion, maybe there is a better way. What do you think?

**Comment by turion** [comment]
Rebased onto https://github.com/NixOS/nixpkgs/pull/120180 to test all agda packages

**Comment by turion** [comment]
@ofborg build agdaPackages.standard-library.passthru.tests

**Comment by turion** [comment]
@ofborg build agdaPackages.standard-library.passthru.tests

**Comment by turion** [comment]
@alexarice @neosimsim what do you think about this? I'd like to merge this before the next standard library is released (in the next days), so we can test the workflow.

**Comment by alexarice** [comment]
I'm not sure about all the changes to the documentation, as I believe that is aimed more at users, but I am happy with all the code changes

**Comment by turion** [comment]
> I'm not sure about all the changes to the documentation, as I believe that is aimed more at users, but I am happy with all the code changes

Yes, I understand. That's why this section is at the end, so mere users will stop reading at some point and not miss anything important. However my impression is that nixpkgs is a rabbit hole, and after some time of using, users automatically evolve into maintainers, and then it's great if they have some docs to get started. (I remember having a hard time starting to contribute to the Haskell ecosystem in nixpkgs, and I'd like to prevent that for future maintainers.) Also, sometimes Agda library authors might want to have their library included, so they can directly jump to that place.

Should I maybe add a sentence saying something to the effect of "If you just want to use Agda, you don't need to read this"?

**Review by jwiegley — APPROVED** [review]

**Comment by turion** [comment]
> I'm not sure about all the changes to the documentation, as I believe that is aimed more at users

For example python also has a section about how python is packaged in nixpkgs. So I think it's fine to have a separate section on that.

**Comment by turion** [comment]
@ofborg build agdaPackages.standard-library.passthru.tests
@ofborg build agdaPackages.passthru.tests

**Comment by turion** [comment]
@ofborg build agda.passthru.tests.all

**Comment by turion** [comment]
@alexarice @neosimsim Comments? Otherwise I'm going to go find someone to merge ;)

File: pkgs/top-level/agda-packages.nix [hunk 2]
```diff
@@ -13,7 +16,10 @@ let
 
     lib = lib.extend (final: prev: import ../build-support/agda/lib.nix { lib = prev; });
 
-    agda = withPackages [] // { inherit withPackages; };
+    agda = withPackages [] // {
+      inherit withPackages;
+      passthru.tests.all = withPackages (lib.filter (pkg: agdaPackages.lib.isUnbrokenAgdaPackage pkg) (lib.attrValues agdaPackages));
```
Comment (alexarice) [inline pkgs/top-level/agda-packages.nix:21]: This looks a little dodgy now that I look at it. I think `self` here should be `agdaPackages` (I could be totally wrong on this)

Comment (turion) [reply pkgs/top-level/agda-packages.nix:21]: Sorry, I can't follow. Where does `self` occur here?

Comment (alexarice) [reply pkgs/top-level/agda-packages.nix:21]: it's further up the file

Comment (alexarice) [reply pkgs/top-level/agda-packages.nix:21]: oh I mistyped, the `agdaPackages` can probably be replaced with `self` (other way round to what I said)

Comment (turion) [reply pkgs/top-level/agda-packages.nix:21]: But this file doesn't have `self`. Maybe you mean a different place?

File: pkgs/build-support/agda/default.nix [hunk 3]
```diff
@@ -77,7 +80,12 @@ let
           find -not \( -path ${everythingFile} -or -path ${lib.interfaceFile everythingFile} \) -and \( ${concatMapStringsSep " -or " (p: "-name '*.${p}'") (extensions ++ extraExtensions)} \) -exec cp -p --parents -t "$out" {} +
           runHook postInstall
         '';
+
         meta = if meta.broken or false then meta // { hydraPlatforms = lib.platforms.none; } else meta;
+
+        # Retrieve all packages from the finished package set that have the current package as a dependency and build them
+        passthru.tests = with builtins;
+          lib.filterAttrs (name: pkg: agdaPackages.lib.isUnbrokenAgdaPackage pkg && elem pname (map (pkg: pkg.pname) pkg.buildInputs)) agdaPackages;
```
Comment (turion) [inline pkgs/build-support/agda/default.nix:88]:
```suggestion
          lib.filterAttrs (name: pkg: self.lib.isUnbrokenAgdaPackage pkg && elem pname (map (pkg: pkg.pname) pkg.buildInputs)) self;
```

@alexarice Is that your suggestion?

File: pkgs/top-level/agda-packages.nix (hunk 2 above)
Comment (alexarice) [reply pkgs/top-level/agda-packages.nix:21]: https://github.com/NixOS/nixpkgs/blob/7c1b01613fd205cdebe459e119b3b702891769c7/pkgs/top-level/agda-packages.nix#L8 here it is

File: pkgs/build-support/agda/default.nix (hunk 3 above)
Comment (alexarice) [reply pkgs/build-support/agda/default.nix:88]: This was not my suggestion but if it works, then we should remove the `agdaPackages` from the arguments and use this instead

Comment (turion) [reply pkgs/build-support/agda/default.nix]: Yes it worked, great improvement :)

File: pkgs/top-level/agda-packages.nix (hunk 2 above)
Comment (turion) [reply pkgs/top-level/agda-packages.nix]: Alright, now I got it thanks!

**Comment by turion** [comment]
@ofborg build agdaPackages.standard-library.passthru.tests
@ofborg build agdaPackages.passthru.tests.allPackages

**Comment by turion** [comment]
@ofborg build agda.passthru.tests.allPackages

**Review by alexarice — APPROVED** [review]

**Comment by Pablito2020** [comment]
@ofborg build python312Packages.insightface.passthru.tests

**Pull request merged**